Fix incorrect HTML in theme & docs; validate HTML in CI

mattxwang opened the pull request:
This PR is motivated from https://github.com/just-the-docs/just-the-docs/pull/1259#issuecomment-1655899503. It adds a new workflow (`CI / Validate HTML (3.1)`) that validates the output of `bundle exec jekyll build`. It does this with two separate tools:

1. The [`html5validator-action`](https://github.com/Cyb3r-Jak3/html5validator-action), which is a wrapper (Docker image + argument forwarding) around the [Nu HTML checker](https://github.com/validator/validator), which is what is used by the [W3C markup validation service](https://validator.w3.org/)
2. [`html-proofer`](https://github.com/gjtorikian/html-proofer), which performs auxiliary checks on the validity of script, image, and link *values*, but not the markup itself
    - note: prior versions of `html-proofer` did use nokogiri to also validate HTML, but the author has elected to remove that feature in versions 4+

I then fix a few issues that are flagged by these tools. I'll split this into,

**changes affecting users**:
- strictly incorrect: in `_layouts/minimal.html`, a `<div>` had duplicate `id`s. I've removed the incorrect one, which is related to...
- semantically wrong (but not technically incorrect): in both `minimal` and `default` layouts, we had two `<div>` tags with `id="main-content-wrap"`. These don't do anything; the associated styling is with the *class* `main-content-wrap`. I've elected to remove these `id`s to avoid confusion and keep the layouts in sync; however, **this is technically a breaking change**
    - observe that `#main-content` is used for the "skip to main content" feature, which I missed in an earlier iteration of this PR

**changes affecting only our documentation**
- a broken link to mermaid docs (I've changed it to a valid one)
- an incorrectly-specified `aux_link` to our own repository
- various links that point to the bare URL `another-page`, which is clearly invalid; I've changed these to point to our homepage
- an incorrect header link
- various links to `http://example.com`, which I've changed to point to our homepage
- an incorrect link to `@flyx`'s profile for the AsciiDoctor gist
- a handful of (otherwise-valid) `http` links that should be `https`: the lunr docs, and patrick's personal website

The commit history shows the Nu validator flagging issues in CI properly in commits [4128b23](https://github.com/just-the-docs/just-the-docs/pull/1305/commits/4128b23ef2e6568fab8df353f820a62ec721a146) and [3527220](https://github.com/just-the-docs/just-the-docs/pull/1305/commits/35272203ba2f485e0d56d2cc78cebe0a31d6f359).

## relevant configuration

- I exclude `github.com` URLs from external link checks with `html-proofer`. This is because GitHub does not like it when we ping too frequently, and rate limits us, which in turn provides many false positives. This is aligned with their documentation, which uses this ignore.
- I've pinned the hash for the 3rd-party action that wraps the W3C markup validation service. This aligns with #1148, but means that we'll have to keep an eye on it for updates.

mattxwang commented:
@pdmosses do you mind taking a look at this when you have a chance? I'm relatively satisfied with this workflow (given that it's already caught another error we introduced in `0.4.0`!).

I did want to specifically get your thoughts on the last errors being flagged by the validator, which is on the [code with line numbers page](https://just-the-docs.com/docs/ui-components/code/line-numbers/); it concerns the [incorrect code example](https://just-the-docs.com/docs/ui-components/code/line-numbers/#examples):

![Screenshot 2023-07-28 at 2 51 09 PM](https://github.com/just-the-docs/just-the-docs/assets/14893287/d5d3846e-6b16-451f-9537-dacdecf331ca)

This ends up generating (understandably) invalid HTML. Should we:

- ignore this file
- remove / change this example to still generate valid HTML

I'm considering the latter, since this was also a point of past confusion for a previous user (I'll have to look for the issue). That being said, I'm not strongly wedded to either approach.

pdmosses commented:
> @pdmosses do you mind taking a look at this when you have a chance? I'm relatively satisfied with this workflow (given that it's already caught another error we introduced in `0.4.0`!).

Thanks for this draft PR. I've skimmed the changes; I'll be able to review the various changes to the theme files, but best to ask another maintainer to check the workflow, as that's out of my comfort zone…

> I did want to specifically get your thoughts on the last errors being flagged by the validator, which is on the [code with line numbers page](https://just-the-docs.com/docs/ui-components/code/line-numbers/); it concerns the [incorrect code example](https://just-the-docs.com/docs/ui-components/code/line-numbers/#examples):
> …
> This ends up generating (understandably) invalid HTML. Should we:
> 
> * ignore this file
> * remove / change this example to still generate valid HTML
> 
> I'm considering the latter, since this was also a point of past confusion for a previous user (I'll have to look for the issue). 

#1227

> The former is easier (and I believe there are some markup issues anyways with `<div>` inside `<code>`). 

The HTML non-conformity reported in [the Jekyll issue](https://github.com/jekyll/jekyll/issues/8411) is the use of a `<table>` inside `<code>`. The `<table>` is produced by Rouge, and valid, but not allowed in phrasing contexts such as `<code>`. (Our `table-wrapper` layout inserts the `<div>`, and doesn't affect validation.)

> That being said, I'm not strongly wedded to either approach.

I've taken a fresh look at the HTML produced by the claimed workarounds. Use of `fix_linenos.html` avoids the poor rendering that arises when `compress-html` meets nested `<pre>` elements, but it does **not** ensure valid HTML! Both the Jekyll `highlight` tag and the kramdown `line_numbers` option with code fences produce `<table>` inside `<code>`.

So those "✅" symbols are somewhat misleading: they don't produce valid HTML with line numbers. I'm in favour of removing the whole page (or at least excluding it from the navigation). We could add a callout at the end of the parent page with a warning that our previously recommended workarounds produce invalid HTML, despite browsers generally managing to render it quite well.

[The Jekyll issue](https://github.com/jekyll/jekyll/issues/8411) is currently still open (and pinned), and AFAIK @ashmaroli still intends to release a plugin to support a new tag for highlighting with line numbers. When that happens, we can update our docs pages accordingly.

max06 commented on .github/workflows/ci.yml:
```diff
@@ -57,6 +57,40 @@ jobs:
+    strategy:
+      fail-fast: false
+      matrix:
+        ruby-version: [3.1]
```
We should add Ruby 3.2 at some point (I'm testing with `ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-linux]`)

mattxwang replied:
Sure; I'll update all of those (+ the template) in lockstep.

max06 commented on .github/workflows/ci.yml:
```diff
@@ -57,6 +57,40 @@ jobs:
+        uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: ${{ matrix.ruby-version }}
+          bundler-cache: false
```
using a cache would be more efficient, wouldn't it?

max06 commented on .github/workflows/ci.yml:
```diff
@@ -57,6 +57,40 @@ jobs:
+          bundler-cache: false
+      - name: Bundle Install
+        run: bundle install
+      - name: Cache HTMLProofer
```
Not sure whats your intention with this.

mattxwang replied:
My mistake, left it in while experimenting.

max06 commented on Gemfile:
```diff
@@ -4,3 +4,5 @@ gemspec
 
 gem "webrick", "~> 1.7"
+
+gem "html-proofer", "~> 5.0", :group => :development
```
There should be a change in the lockfile?

mattxwang replied:
Sorry, a bit confused - what lockfile? We haven't merged #1112 yet.

pdmosses replied:
> We haven't merged #1112 yet.

Wouldn't it be good to do that a.s.a.p.? I agree that a 2nd review is needed – especially in view of my low familiarity with the workings of Bundler, Jekyll, and gem files!

BTW, the GH Pages version of `just-the-docs-tests` is now built with Jekyll 4, but its `Gemfile` supports local building with Jekyll 3 or 4. (It doesn't need the legacy `github-pages` gem, so I never tried to support that.)

max06 replied:
Adding a gem to the Gemfile and running bundle install on your machine causes a change to the lockfile. These should be committed together with the change to the Gemfile. Otherwise each and every pipeline run (without cache) will detect the difference between these 2 files, install the locked dependencies from the lockfile and any current version (respecting the constraints) of the new gem.

The `Gemfile` just contains a list of gems and constraints. The `Gemfile.lock` contains a resolved dependency tree. `bundle install` will take the resolved tree from the lockfile and make sure the `Gemfile` hasn't changed.

max06 replied:
Oh well, nevermind. We should add the lockfile asap.

mattxwang replied:
Definitely agree that we should commit a lockfile (and ideally, lockfiles for our test fixtures as well); I think #1112 should be the avenue for that (I can handle merging these in whatever order we'd prefer).

pdmosses replied:
See also [my reference to the updated Jekyll recommendation](https://github.com/just-the-docs/just-the-docs/pull/1112#discussion_r1063390196).

max06 replied:
Let's get #1307 merged (yes, new one), then rebase this PR.

mattxwang replied:
Hm, #1307 is requiring some back and forth. To better understand - why is it necessary to merge in that *before* this PR, and why not #1112? From my perspective, I don't see a direct issue that interferes with the purpose of this PR. 

For what it's worth: after this PR, I'll cut a `v0.5.x` and move forward on `v0.6` with the fixed navigation.

ashmaroli commented:
As an FYI: I won't be releasing the plugin mentioned in this thread. Apologies for any misgivings.

mattxwang commented:
Thanks @pdmosses and @max06 for your quick reviews! I think I've addressed most of the issues re: the workflow.

In terms of 

> I'm in favour of removing the whole page (or at least excluding it from the navigation). We could add a callout at the end of the parent page with a warning that our previously recommended workarounds produce invalid HTML, despite browsers generally managing to render it quite well.

Sounds good to me! If you don't mind, I'll punt this to another issue (primarily so we can release our bugfixes relatively quickly). We can then workshop how to best demonstrate code with line numbers without misleading users about the validity of our HTML.

With that in mind, opening this up for a more in-depth review; I think after merging this (since it addresses another markup issue), I can cut a new release.

mattxwang commented:
Have now opened #1306 re: removing/editing code-with-line-numbers page.

mattxwang commented:
Going to make a judgement call here and merge this PR in, mainly because:

- I'd like to release soon, especially since this unblocks me to work on merging one of our big asks (likely light mode) in first-party
- it seems like we're in a bit of a review glut
- I'm quite confident in the changes + other than fixes, this should have little end-user impact

Would we like to clarify our stance on how maintainer PRs are reviewed & merged? In an ideal world, we'd treat maintainer work like any other third-party contribution, but I also think we're quite a bit more resource-strapped than other repos (ex Stylelint). I know I've frequently self-merged PRs (especially ones that aren't user-facing), not sure how we feel about that.
